pkg/storaged/client.js

```diff
@@ -206,6 +214,203 @@
+            } catch (err) {
+                console.error(`btrfs filesystem show ${uuid}`, err);
+                return;
+            }
+            for (const line of usage_output.split("\n")) {
+                const match = usage_regex.exec(line);
+                if (match) {
+                    const { used, device } = match.groups;
+                    usages[device] = used;
+                }
+            }
+            uuids_usage[uuid] = usages;
+        } else {
+            uuids_subvols[uuid] = null;
+            uuids_usage[uuid] = null;
+        }
+    }
+
+    if (!deep_equal(client.uuids_btrfs_subvols, uuids_subvols) || !deep_equal(client.uuids_btrfs_usage, uuids_usage) ||
+        !deep_equal(client.uuids_btrfs_default_subvol, btrfs_default_subvol)) {
+        debug("btrfs_pol new subvols:", uuids_subvols);
+        client.uuids_btrfs_subvols = uuids_subvols;
+        client.uuids_btrfs_usage = uuids_usage;
+        debug("btrfs_pol usage:", uuids_usage);
+        client.uuids_btrfs_default_subvol = btrfs_default_subvol;
+        debug("btrfs_pol default subvolumes:", btrfs_default_subvol);
+        client.update();
+    }
+}
+
+function btrfs_findmnt_poll() {
+    if (!client.btrfs_mounts)
+        client.btrfs_mounts = { };
+
+    const update_btrfs_mounts = output => {
+        const btrfs_mounts = {};
+        try {
+            // Extract the data into a { uuid: { subvolid: { subvol, target } } }
+            const mounts = JSON.parse(output);
+            if ("filesystems" in mounts) {
+                for (const fs of mounts.filesystems) {
+                    const subvolid_match = fs.options.match(/subvolid=(?<subvolid>\d+)/);
+                    const subvol_match = fs.options.match(/subvol=(?<subvol>[\w\\/]+)/);
+                    if (!subvolid_match && !subvol_match) {
+                        console.warn("findmnt entry without subvol and subvolid", fs);
+                        break;
+                    }
+
+                    const { subvolid } = subvolid_match.groups;
+                    const { subvol } = subvol_match.groups;
+                    const subvolume = {
+                        pathname: subvol,
+                        id: subvolid,
+                        mount_points: [fs.target],
+                    };
+
+                    if (!(fs.uuid in btrfs_mounts)) {
+                        btrfs_mounts[fs.uuid] = { };
+                    }
+
+                    // We need to handle multiple mounts, they are listed seperate.
+                    if (subvolid in btrfs_mounts[fs.uuid]) {
+                        btrfs_mounts[fs.uuid][subvolid].mount_points.push(fs.target);
+                    } else {
+                        btrfs_mounts[fs.uuid][subvolid] = subvolume;
+                    }
+                }
+            }
+        } catch (exc) {
+            if (exc.message)
+                console.error("unable to parse findmnt JSON output", exc);
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20240111-133844-24a81545-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)